onnxruntime/core/platform/windows/logging/etw_sink.cc

```diff
@@ -137,28 +137,47 @@ void NTAPI EtwRegistrationManager::ORT_TL_EtwEnableCallback(
 EtwRegistrationManager::~EtwRegistrationManager() {
   std::lock_guard<OrtMutex> lock(callbacks_mutex_);
   callbacks_.clear();
-  ::TraceLoggingUnregister(etw_provider_handle);
+  if (initialized_ || initializing_) {
+    std::lock_guard<OrtMutex> init_lock(init_mutex_);
+      assert(!initializing_);
+      if (initialized_) {
+        ::TraceLoggingUnregister(etw_provider_handle);
+        initialized_ = false;
+      }
+  }
 }
 
 EtwRegistrationManager::EtwRegistrationManager() {
 }
 
-void EtwRegistrationManager::LazyInitialize() {
-  if (!initialized_) {
+void EtwRegistrationManager::LazyInitialize() try {
+  if (!initialized_ && !initializing_) {
     std::lock_guard<OrtMutex> lock(init_mutex_);
-    if (!initialized_) {  // Double-check locking pattern
-      initialized_ = true;
+    if (!initialized_ && !initializing_) {  // Double-check locking pattern
```

Comment:
I think that the need to check the same condition before and after locking the mutex can be avoided by using an atomic compare-and-exchange operation like [InterlockedCompareExchange](https://learn.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedcompareexchange). In fact, this may avoid the need for a mutex at all.

Before:
```c++
if (state_ == NotInitialized) {
  std::lock_guard<OrtMutex> lock(init_mutex_);
  if (state_ == NotIntialized) {
    // ..
  }
}
```

After:
```c++
const State initial_state = state;  // Copy it since state_ could be modified by another thread later.
if (initial_state != State::NotInitialized) {
  return;
}

const State desired_state = State::Initializing;
const State result = InterlockedCompareExchange(reinterpret_cast<LONG volatile *>(&state_), // Memory to modify atomically
                                                static_cast<LONG>(desired_state),  // Value to set if successful
                                                static_cast<LONG>(initial_state));  // Value to compare with memory

if (result == initial_state) {
  // Atomic compare-exchange was successful. Only one thread will succeed.
  // state_ is now equal to State::Initializing.

  try {
    // Do stuff to initialize: ::TraceLoggingRegisterEx(...)
    state_ = State::Initialized;
  } catch (...) {
    // Something bad happened
    state_ = State::Failed;
  }
}

```

Something like that.

Reply:
I think in general a mutex is needed because the destructor should wait for initialization to complete if it gets triggered during "initializing."